RecoHGCal/TICL/plugins/TrackstersProducer.cc

```diff
@@ -0,0 +1,90 @@
+    : myAlgo_(std::make_unique<PatternRecognitionbyCA>(ps)) {
+  clusters_token_ = consumes<std::vector<reco::CaloCluster>>(
+      ps.getParameter<edm::InputTag>("hgcal_layerclusters"));
+  filtered_layerclusters_mask_token_ = consumes<std::vector<std::pair<unsigned int, float>>>(
+      ps.getParameter<edm::InputTag>("filtered_layerclusters_mask"));
+  original_layerclusters_mask_token_ =
+      consumes<std::vector<float>>(ps.getParameter<edm::InputTag>("original_layerclusters_mask"));
+  produces<std::vector<Trackster>>("TrackstersByCA");
+  produces<std::vector<float>>();  // Mask to be applied at the next iteration
+}
+
+void TrackstersProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // hgcalMultiClusters
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("hgcal_layerclusters", edm::InputTag("hgcalLayerClusters"));
+  desc.add<edm::InputTag>("filtered_layerclusters_mask",
+                          edm::InputTag("FilteredLayerClusters", "iterationLabelGoesHere"));
+  desc.add<edm::InputTag>("original_layerclusters_mask",
+                          edm::InputTag("hgcalLayerClusters", "InitialLayerClustersMask"));
+  desc.add<int>("algo_verbosity", 0);
+  desc.add<double>("min_cos_theta", 0.915);
+  desc.add<double>("min_cos_pointing", -1.);
+  desc.add<int>("missing_layers", 0);
+  desc.add<int>("min_clusters_per_ntuplet", 10);
+  descriptions.add("Tracksters", desc);
+}
+
+void TrackstersProducer::produce(edm::Event& evt, const edm::EventSetup& es) {
+  auto result = std::make_unique<std::vector<Trackster>>();
+  auto output_mask = std::make_unique<std::vector<float>>();
+
+  edm::Handle<std::vector<reco::CaloCluster>> cluster_h;
+  edm::Handle<std::vector<std::pair<unsigned int, float>>> filtered_layerclusters_mask_h;
+  edm::Handle<std::vector<float>> original_layerclusters_mask_h;
+
+  evt.getByToken(clusters_token_, cluster_h);
+  evt.getByToken(filtered_layerclusters_mask_token_, filtered_layerclusters_mask_h);
+  evt.getByToken(original_layerclusters_mask_token_, original_layerclusters_mask_h);
+  std::cout << "TrackstersProducer::produce" << std::endl;
+
+  const auto& layerClusters = *cluster_h;
+  const auto& inputClusterMask = *filtered_layerclusters_mask_h;
+  std::unique_ptr<std::vector<std::pair<unsigned int, float>>> filteredLayerClusters;
+  myAlgo_->makeTracksters(evt, es, layerClusters, inputClusterMask, *result);
+
+  // Now update the global mask and put it into the event
+  output_mask->reserve(original_layerclusters_mask_h->size());
+  // Copy over the previous state
+  std::copy(std::begin(*original_layerclusters_mask_h), std::end(*original_layerclusters_mask_h),
+            std::back_inserter(*output_mask));
+  // Mask the used elements, accordingly
+  for (auto const& trackster : *result) {
+    for (auto const v : trackster.vertices) {
+      // TODO(rovere): for the moment we mask the layer cluster completely. In
```

Comment:
should this todo be implemented before the PR is merged?

Reply:
Same as other TODO, but on an even longer timescale.